### PR TITLE
IAP: bug fixes and cleanup

### PR DESCRIPTION
**Pull request by sbdt-google**
Several bug fixes for IAP
General cleanup of the IAP module (aggregate_interference cleanup will follow):
  - ESC data extraction done on single place in common/data.py (so will be reused in aggregate_interference and will avoid code duplication)
  - FSS parameter extraction was wrong for tilt. Remove duplication by using the common/data.py routine already used elsewhere
  - building the final post-IAP dictionary was wrong: the longitude was constantly erased and only one was kept per latitude (note that this expains some issues see in KeyError exception running MCP)
  - general cleanup to use the common naming for variables, and improve the readability of the main IAP loop

### CODE REVIEW

File: src/harness/reference_models/iap/iap.py [hunk 1]
```diff
@@ -87,365 +86,370 @@
...
+    fairshare_channel = 0
+    if num_unsatisfied_grants_channel > 0:
+      fairshare_channel = float(iap_threshold) / num_unsatisfied_grants_channel
+
+    num_unsatisfied_grants_channels.append(num_unsatisfied_grants_channel)
+    iap_threshold_channels.append(iap_threshold)
+    fairshare_channels.append(fairshare_channel)
+
+  # List of grants initialized with max EIRP
+  grants_eirp = [grant.max_eirp for grant in neighbor_grants]
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: no change in the loop, just indentation changed by transforming the if above

File: src/harness/reference_models/iap/iap.py [hunk 2]
```diff
@@ -87,365 +86,370 @@
...
+      if not grants_satisfied[gr_idx]:
+        continue
+      if num_unsatisfied_grants > 0:
+        num_unsatisfied_grants -= 1
+      for ch_idx, channel in enumerate(channels):
+        # Grant interferes with protection point over channel
+        if grants_overlap_flag[(gr_idx, ch_idx)]:
+          iap_threshold_channels[ch_idx] = (iap_threshold_channels[ch_idx] -
+                                            grants_interference[(gr_idx, ch_idx)])
+          num_unsatisfied_grants_channels[ch_idx] -= 1
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: use += or -= instead for clarity. Here and elsewhere

File: src/harness/reference_models/iap/iap.py [hunk 3]
```diff
@@ -87,365 +86,370 @@
...
+                                        + interf.IN_BAND_INSERTION_LOSS)
 
-  grant_objects = data.getGrantObjectsFromFAD(sas_uut_fad_object, sas_th_fad_objects)
+  grants = data.getGrantObjectsFromFAD(sas_uut_fad_object, sas_th_fad_objects)
 
-  # Get number of SAS 
+  # Get number of SAS
   num_sas = len(sas_th_fad_objects) + 1
 
-  # Get the protection point of the ESC
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: replaced this by the common/data.py getEscInfo routine

File: src/harness/reference_models/iap/iap.py [hunk 4]
```diff
@@ -87,365 +86,370 @@
...
+  grants = data.getGrantObjectsFromFAD(sas_uut_fad_object, sas_th_fad_objects)
 
-  # Get number of SAS 
+  # Get number of SAS
   num_sas = len(sas_th_fad_objects) + 1
 
   logging.debug('$$$$ Getting GRID points for GWPZ Protection Area $$$$')
   # Get Fine Grid Points for a GWPZ protection area
-  protection_points = utils.GridPolygon(protected_entity['zone']['features'][0]['geometry'],
+  protection_points = utils.GridPolygon(gwpz_record['zone']['features'][0]['geometry'],
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: changed the naming of protection_entity to make it clear what we are processing

File: src/harness/reference_models/iap/iap.py [hunk 5]
```diff
@@ -459,190 +463,192 @@ def performIapForPpa(protected_entity, sas_uut_fad_object, sas_th_fad_objects,
...
-      ['operationParam']['operationFrequencyRange']
-  fss_low_freq = fss_freq_range['lowFrequency']
-  fss_high_freq = fss_freq_range['highFrequency']
-
-  # Get FSS information
-  fss_info = data.FssInformation(height_agl=fss_point['height'],
-               max_gain_dbi=fss_point['antennaGain'],
-               pointing_azimuth=fss_point['antennaAzimuth'],
-               pointing_elevation=fss_point['antennaDowntilt'])
+  # Get the FSS infos
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: used the common/data routine. This also suppress bugs:
 - elevation was wrong (lack the minus)
 - handle the AMSL case

File: src/harness/reference_models/iap/iap.py [hunk 6]
```diff
@@ -459,190 +463,192 @@ def performIapForPpa(protected_entity, sas_uut_fad_object, sas_th_fad_objects,
...
 
   # Get FSS T&C Flag value
-  fss_ttc_flag = protected_entity['ttc']
+  fss_ttc_flag = fss_record['ttc']
 
-  # Get the protection point of the FSS
-  fss_point = protected_entity['record']['deploymentParam'][0]['installationParam']
-  protection_point = (fss_point['longitude'], fss_point['latitude'])
+  # Get the FSS infos
+  fss_point, fss_info, fss_freq_range = data.getFssInfo(fss_record)
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap.py]: use the common/data getFssInfo routine. Suppress the 2 bugs mentioned before.

File: src/harness/reference_models/iap/iap_example.py [hunk 7]
```diff
@@ -82,19 +82,19 @@
 
   sas_uut = FullActivityDump({
             'cbsd': [cbsd_0, cbsd_1, cbsd_2, cbsd_3, cbsd_4, cbsd_5, cbsd_6,cbsd_7, cbsd_8, cbsd_9, cbsd_10, cbsd_11, cbsd_12, cbsd_13, cbsd_14, cbsd_15, cbsd_16, cbsd_17, cbsd_18, cbsd_19],
```
Comment (sbdt-google) [inline src/harness/reference_models/iap/iap_example.py]: just changed the indentation in this file

**Comment by sbdt-google** [comment]
Just trying to fix bugs in the code, and make the code in a state where we
can more easily develop unit tests.
If you look at the old vs new code in 2 windows, you will see that there
are no real changes for the "cleanup" part, and it should be quite fast to
review. It looks big
because indentation change gives the impression of a lot of changes in
GitHub.

I urge you to speed up this review and dedicate someone to do it asap,
because  MCP cannot be run and fully tested without some of these bug
fixes.

I am afraid that given the pace of fixes and tests done by the community
during the V&V, we may need to extend that V&V period:

 - the fixed bug means that noone has gone through the steps of running the
full MCP on their own SAS (otherwise the bugs would have been
reported/fixed before).
 - the various code related to IAP, interference and pre-iap still lack all
unit testing (except for the aggregate_interference module). It was
understood several weeks ago that we would let
   the code be submitted despite that.. And that these unit test would be
developed in a second step.
   But  none of them have been developed... We only have 'examples' which
tests that the code run without crash but no test that the most complex
part of the logic is ok (especially troublesome for
   IAP and pre-IAP stuff).

=> from my point of view, we need to do an extra effort to find and fix the
remaining issues in the code, by both running the test harness against the
SAS and developing some unit tests.
I am willing to help in a fast pace in next few days fixing the missing
unit test and discovered bugs, but I also need responsiveness in the review
process.

Thanks
serge

**Comment by jni2000** [comment]
thanks, Serge. We are in process right now and will get back to you as soon as possible. The not late than tomorrow morning is my conservative estimation as we do not know if we will see any issues or not. We are targeting to get back to you in hours if everything goes on smooth.

**Comment by sbdt-google** [comment]
Thanks, submitting now

**Pull request merged**